hs/src/Reach/Optimize.hs

```diff
@@ -246,6 +248,99 @@ opt_v2v v = fst <$> opt_v2p v
 opt_v2a :: DLVar -> App DLArg
 opt_v2a v = snd <$> opt_v2p v
 
+class IsCom a where
+  isCom :: a -> Maybe (DLStmt, a)
+
+instance IsCom DLTail where
+  isCom = \case
+    DT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom CTail where
+  isCom = \case
+    CT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom ETail where
+  isCom = \case
+    ET_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLStep where
+  isCom = \case
+    LLS_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLConsensus where
+  isCom = \case
+    LLC_Com m k -> Just (m, k)
+    _ -> Nothing
+
+class LiftPure a where
+  -- Returns what `DLVar` is `DL_Set` to and the pure `DLStmt`s along the way
+  -- Returns `Nothing` if there are impure stmts in `a` or there is no `DL_Set`
+  liftPure :: DLVar -> a -> (Maybe DLArg, [DLStmt])
```

Comment:
This is a bit of bike-shedding and a demand for purity... I think `DLVar -> a -> Maybe (DLArg, [DLStmt])` is a more accurate type.